### PR TITLE
Qmc5883

### PR DESCRIPTION
**Pull request by osgjps**
## Description:

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [ X] The pull request is done against the latest development branch
  - [ X] Only relevant files were touched
  - [ X] Only one feature/fix was added per PR and the code change compiles without warnings
  - [ X The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [ X] The code change is tested and works with Tasmota core ESP32 V.2.0.2.1
  - [ X] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

**Comment by osgjps** [comment]
This PR is technically for the QMC5883.  I don't have an HMC5883L to test with, I misnamed the branch when I created it.

**Comment by ascillato2** [comment]
Hi, your PR will fail for all languages versions except for English. Please, add the missing keys to the rest of language files. Don't need to translate it. Just copy the same you have in the English file.

File: tasmota/i18n.h [hunk 1]
```diff
@@ -122,6 +122,11 @@
 #define D_JSON_MODEL "Model"
 #define D_JSON_MOISTURE "Moisture"
 #define D_JSON_MQTT_COUNT "MqttCount"
+// xsns_96_QMC5883.ino
+#define D_JSON_MX "Compass  X-Axis"
+#define D_JSON_MY "Compass  Y-Axis"
+#define D_JSON_MZ "Compass Z-Axis"
+#define D_JSON_HEADING "Compass Heading"
```
Comment (barbudor) [inline tasmota/i18n.h:129]: These are wrong names for JSON keys
A JSON key must be a valid variable name in Javascript with only alphanumeric characters  + underscore and starting with an alpha character.
Spaces, dash or other separators are illegal.
Please fixe

File: tasmota/xsns_96_qmc5883.ino [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+{
+int   MX;
+int   MY;
+int   MZ;
+int   HG;
+uint8_t   status;
+bool      ready;
+uint8_t   i2c_address;
+
+} QMC5883;
```
Comment (barbudor) [inline tasmota/xsns_96_qmc5883.ino:66]: Do not allocate statically the memory for your driver
This is taking useless amount of RAM when the driver is part of a build but not used

Use dynamic allocation as in the HMC5883 driver

File: tasmota/xsns_96_qmc5883.ino [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+bool Xsns96(byte function)
+{
+  if (!I2cEnabled(XI2C_66)) { return false; }
+
+  bool result = false;
+
+  if (FUNC_INIT == function) {
+    QMC5883_Init();
+  }
+  else if (QMC5883.ready) {
```
Comment (barbudor) [inline tasmota/xsns_96_qmc5883.ino:202]: Similar, look at the HMC5883 driver. Use the dynamic pointer for that purpose instead of wasting a byte

**Review by barbudor — CHANGES_REQUESTED** [review]
Changes to be done to be a good Tasmota driver

**Pull request closed without merging**